File: structy/BinaryTree/level_averages.py
class Node:
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None

a = Node(3)
b = Node(11)
c = Node(4)
d = Node(4)
e = Node(-2)
f = Node(1)

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#       3
#    /    \
#   11     4
#  / \      \
# 4   -2     1
from statistics import mean

def level_averages(root):
    if not root:
        return []

    stack = [ (root, 0) ]
    levels = []

    while stack:
        current, level = stack.pop()

        if len(levels) == level:
            levels.append([ current.val ])
        else:
            levels[level].append(current.val)

        if current.right:
            stack.append( (current.right, level+1) )
        if current.left:
            stack.append( (current.left, level+1) )

    averages = list(map(lambda x: mean(x), levels))
    return averages

print(level_averages(a)) # -> [ 3, 7.5, 1 ] 

File: structy/BinaryTree/levelAverages.js
function levelAverages(root) {
    const levels = [];
    fillLevels(root, levels, 0);
    const averages = levels.map(level =>
        average(level));

    return averages
};

function fillLevels(root, levels, levelNum) {
    if (!root) return;

    if (levels.length === levelNum) {
        levels.push([ root.val ])
    } else {
        levels[levelNum].push(root.val)
    };

    fillLevels(root.left, levels, levelNum+1)
    fillLevels(root.right, levels, levelNum+1)
};

function average(array) {
    return array.reduce((a, b) => a +b) / array.length
};

class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const a = new Node(3);
const b = new Node(11);
const c = new Node(4);
const d = new Node(4);
const e = new Node(-2);
const f = new Node(1);

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//       3
//    /    \
//   11     4
//  / \      \
// 4   -2     1

console.log(levelAverages(a)); // -> [ 3, 7.5, 1 ] 
